File: CS/DXWebApplication1/AzureFileStorageProvider.cs
using DevExpress.Web;
using Microsoft.WindowsAzure.Storage;
using Microsoft.WindowsAzure.Storage.Auth;
using Microsoft.WindowsAzure.Storage.File;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace DXWebApplication1 {
	public class AzureFileStorageProvider : FileSystemProviderBase {
		const string accountName = "YourAccountName";
		const string keyValue = "YourKeyValue";
		const string shareName = "YourShareName";
		StorageCredentials cred = new Microsoft.WindowsAzure.Storage.Auth.StorageCredentials(accountName, keyValue);
		CloudStorageAccount account;
		CloudFileClient client;
		CloudFileShare fileShare;
		CloudFileDirectory rootDirectory;
		string rootFolderDisplayName = "root";
		public override string RootFolderDisplayName { get { return rootFolderDisplayName; } }
		public AzureFileStorageProvider(string rootFolder) : base(rootFolder) {
			account = new CloudStorageAccount(cred, true);
			client = account.CreateCloudFileClient();
			fileShare = client.GetShareReference(shareName);
			rootDirectory = fileShare.GetRootDirectoryReference();
		}
		public override IEnumerable<FileManagerFile> GetFiles(FileManagerFolder folder) {
			String prefix = folder.FullName;

			return GetDirectory(prefix)
			 .ListFilesAndDirectories()
			 .OfType<CloudFile>()
			 .Select(cf => new FileManagerFile(this, folder, cf.Name))
			 .ToList();
		}
		public override IEnumerable<FileManagerFolder> GetFolders(FileManagerFolder parentFolder) {
			String prefix = parentFolder.FullName;

			return GetDirectory(prefix)
				.ListFilesAndDirectories()
				.OfType<CloudFileDirectory>()
				.Select(cfd => new FileManagerFolder(this, parentFolder, cfd.Name))
				.ToList();
		}
		public override bool Exists(FileManagerFile file) {
			return GetFile(file).Exists();
		}
		public override bool Exists(FileManagerFolder folder) {
			return GetDirectory(folder.FullName).Exists();
		}
		public override System.IO.Stream ReadFile(FileManagerFile file) {
			var fileToRead = GetFile(file);
			return fileToRead.OpenRead();
		}
		public override void CopyFile(FileManagerFile file, FileManagerFolder newParentFolder) {
			var sourceFile = GetFile(file);
			GetFile(newParentFolder.FullName, file.Name).StartCopy(sourceFile);
		}
		public override void CopyFolder(FileManagerFolder folder, FileManagerFolder newParentFolder) {
			var sourceFolder = GetDirectory(folder.FullName);
			CopyFolderInternal(GetDirectory(folder.FullName), GetDirectory(newParentFolder.FullName), folder.Name);
		}
		public override void CreateFolder(FileManagerFolder parent, string name) {
			GetDirectory(parent.FullName).GetDirectoryReference(name).CreateIfNotExists();
		}
		public override void DeleteFile(FileManagerFile file) {
			GetFile(file).DeleteIfExists();
		}
		public override void DeleteFolder(FileManagerFolder folder) {
			DeleteFolderInternal(GetDirectory(folder.FullName));
		}
		public override DateTime GetLastWriteTime(FileManagerFile file) {
			var shareFile = GetFile(file);
			shareFile.FetchAttributes();
			var properties = shareFile.Properties;
			return (properties.LastModified == null) ? base.GetLastWriteTime(file) : Convert.ToDateTime(properties.LastModified.Value.DateTime);
		}
		public override DateTime GetLastWriteTime(FileManagerFolder folder) {
			var directory = GetDirectory(folder.FullName);
			directory.FetchAttributes();
			var properties = directory.Properties;
			return (properties.LastModified == null) ? base.GetLastWriteTime(folder) : Convert.ToDateTime(properties.LastModified.Value.DateTime);
		}
		public override long GetLength(FileManagerFile file) {
			var shareFile = GetFile(file);
			shareFile.FetchAttributes();
			var properties = shareFile.Properties;
			return (properties.Length == -1) ? base.GetLength(file) : properties.Length;
		}
		public override void MoveFile(FileManagerFile file, FileManagerFolder newParentFolder) {
			GetFile(newParentFolder.FullName, file.Name).StartCopy(GetFile(file));
			GetFile(file).DeleteIfExists();
		}
		public override void MoveFolder(FileManagerFolder folder, FileManagerFolder newParentFolder) {
			CopyFolder(folder, newParentFolder);
			DeleteFolder(folder);
		}
		public override void RenameFile(FileManagerFile file, string name) {
			GetFile(file.Folder.FullName, name).StartCopy(GetFile(file));
			GetFile(file).DeleteIfExists();
		}
		public override void RenameFolder(FileManagerFolder folder, string name) {
			CopyFolderInternal(folder, name);
			DeleteFolder(folder);
		}
		public override void UploadFile(FileManagerFolder folder, string fileName, Stream content) {
			GetFile(folder.FullName, fileName).UploadFromStream(content);
		}
		CloudFileDirectory GetDirectory(string prefix) {
			return (prefix == "") ? rootDirectory : rootDirectory.GetDirectoryReference(prefix);
		}
		CloudFile GetFile(FileManagerFile file) {
			return GetDirectory(file.Folder.FullName).GetFileReference(file.Name);
		}
		CloudFile GetFile(string folderName, string fileName) {
			return GetDirectory(folderName).GetFileReference(fileName);
		}
		void CopyFolderInternal(CloudFileDirectory sourceFolder, CloudFileDirectory targetFolder, string directoryName) {
			targetFolder.GetDirectoryReference(directoryName).CreateIfNotExists();
			var newFolder = targetFolder.GetDirectoryReference(directoryName);
			var items = sourceFolder.ListFilesAndDirectories();
			foreach(IListFileItem item in items) {
				if(item.GetType() == typeof(CloudFile)) {
					var file = (CloudFile)item;
					newFolder.GetFileReference(file.Name).StartCopy(file);
				}
				else {
					var directory = (CloudFileDirectory)item;
					CopyFolderInternal(directory, newFolder, directory.Name);
				}
			}
		}
		void CopyFolderInternal(FileManagerFolder sourceFolder, string newName) {
			CopyFolderInternal(GetDirectory(sourceFolder.FullName), GetDirectory(sourceFolder.Parent.FullName), newName);
		}
		void DeleteFolderInternal(CloudFileDirectory sourceFolder) {
			var items = sourceFolder.ListFilesAndDirectories().ToList();
			if(items.Count == 0) {
				sourceFolder.DeleteIfExists();
				return;
			}
			foreach(IListFileItem item in items) {
				if(item.GetType() == typeof(CloudFile)) {
					var file = (CloudFile)item;
					file.DeleteIfExists();
				}
				else {
					var directory = (CloudFileDirectory)item;
					DeleteFolderInternal(directory);
				}
			}
			sourceFolder.DeleteIfExists();
		}
	}
}

File: VB/DXWebApplication1/AzureFileStorageProvider.vb
Imports DevExpress.Web
Imports Microsoft.WindowsAzure.Storage
Imports Microsoft.WindowsAzure.Storage.Auth
Imports Microsoft.WindowsAzure.Storage.File
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq

Namespace DXWebApplication1

    Public Class AzureFileStorageProvider
        Inherits FileSystemProviderBase

        Const accountName As String = "YourAccountName"

        Const keyValue As String = "YourKeyValue"

        Const shareName As String = "YourShareName"

        Private cred As StorageCredentials = New StorageCredentials(accountName, keyValue)

        Private account As CloudStorageAccount

        Private client As CloudFileClient

        Private fileShare As CloudFileShare

        Private rootDirectory As CloudFileDirectory

        Private rootFolderDisplayNameField As String = "root"

        Public Overrides ReadOnly Property RootFolderDisplayName As String
            Get
                Return rootFolderDisplayNameField
            End Get
        End Property

        Public Sub New(ByVal rootFolder As String)
            MyBase.New(rootFolder)
            account = New CloudStorageAccount(cred, True)
            client = account.CreateCloudFileClient()
            fileShare = client.GetShareReference(shareName)
            rootDirectory = fileShare.GetRootDirectoryReference()
        End Sub

        Public Overrides Function GetFiles(ByVal folder As FileManagerFolder) As IEnumerable(Of FileManagerFile)
            Dim prefix As String = folder.FullName
            Return GetDirectory(prefix).ListFilesAndDirectories().OfType(Of CloudFile)().[Select](Function(cf) New FileManagerFile(Me, folder, cf.Name)).ToList()
        End Function

        Public Overrides Function GetFolders(ByVal parentFolder As FileManagerFolder) As IEnumerable(Of FileManagerFolder)
            Dim prefix As String = parentFolder.FullName
            Return GetDirectory(prefix).ListFilesAndDirectories().OfType(Of CloudFileDirectory)().[Select](Function(cfd) New FileManagerFolder(Me, parentFolder, cfd.Name)).ToList()
        End Function

        Public Overrides Function Exists(ByVal file As FileManagerFile) As Boolean
            Return GetFile(file).Exists()
        End Function

        Public Overrides Function Exists(ByVal folder As FileManagerFolder) As Boolean
            Return GetDirectory(folder.FullName).Exists()
        End Function

        Public Overrides Function ReadFile(ByVal file As FileManagerFile) As Stream
            Dim fileToRead = GetFile(file)
            Return fileToRead.OpenRead()
        End Function

        Public Overrides Sub CopyFile(ByVal file As FileManagerFile, ByVal newParentFolder As FileManagerFolder)
            Dim sourceFile = GetFile(file)
            GetFile(newParentFolder.FullName, file.Name).StartCopy(sourceFile)
        End Sub

        Public Overrides Sub CopyFolder(ByVal folder As FileManagerFolder, ByVal newParentFolder As FileManagerFolder)
            Dim sourceFolder = GetDirectory(folder.FullName)
            CopyFolderInternal(GetDirectory(folder.FullName), GetDirectory(newParentFolder.FullName), folder.Name)
        End Sub

        Public Overrides Sub CreateFolder(ByVal parent As FileManagerFolder, ByVal name As String)
            GetDirectory(parent.FullName).GetDirectoryReference(name).CreateIfNotExists()
        End Sub

        Public Overrides Sub DeleteFile(ByVal file As FileManagerFile)
            GetFile(file).DeleteIfExists()
        End Sub

        Public Overrides Sub DeleteFolder(ByVal folder As FileManagerFolder)
            DeleteFolderInternal(GetDirectory(folder.FullName))
        End Sub

        Public Overrides Function GetLastWriteTime(ByVal file As FileManagerFile) As Date
            Dim shareFile = GetFile(file)
            shareFile.FetchAttributes()
            Dim properties = shareFile.Properties
            Return If(properties.LastModified Is Nothing, MyBase.GetLastWriteTime(file), Convert.ToDateTime(properties.LastModified.Value.DateTime))
        End Function

        Public Overrides Function GetLastWriteTime(ByVal folder As FileManagerFolder) As Date
            Dim directory = GetDirectory(folder.FullName)
            directory.FetchAttributes()
            Dim properties = directory.Properties
            Return If(properties.LastModified Is Nothing, MyBase.GetLastWriteTime(folder), Convert.ToDateTime(properties.LastModified.Value.DateTime))
        End Function

        Public Overrides Function GetLength(ByVal file As FileManagerFile) As Long
            Dim shareFile = GetFile(file)
            shareFile.FetchAttributes()
            Dim properties = shareFile.Properties
            Return If(properties.Length = -1, MyBase.GetLength(file), properties.Length)
        End Function

        Public Overrides Sub MoveFile(ByVal file As FileManagerFile, ByVal newParentFolder As FileManagerFolder)
            GetFile(newParentFolder.FullName, file.Name).StartCopy(GetFile(file))
            GetFile(file).DeleteIfExists()
        End Sub

        Public Overrides Sub MoveFolder(ByVal folder As FileManagerFolder, ByVal newParentFolder As FileManagerFolder)
            CopyFolder(folder, newParentFolder)
            DeleteFolder(folder)
        End Sub

        Public Overrides Sub RenameFile(ByVal file As FileManagerFile, ByVal name As String)
            GetFile(file.Folder.FullName, name).StartCopy(GetFile(file))
            GetFile(file).DeleteIfExists()
        End Sub

        Public Overrides Sub RenameFolder(ByVal folder As FileManagerFolder, ByVal name As String)
            CopyFolderInternal(folder, name)
            DeleteFolder(folder)
        End Sub

        Public Overrides Sub UploadFile(ByVal folder As FileManagerFolder, ByVal fileName As String, ByVal content As Stream)
            GetFile(folder.FullName, fileName).UploadFromStream(content)
        End Sub

        Private Function GetDirectory(ByVal prefix As String) As CloudFileDirectory
            Return If(Equals(prefix, ""), rootDirectory, rootDirectory.GetDirectoryReference(prefix))
        End Function

        Private Function GetFile(ByVal file As FileManagerFile) As CloudFile
            Return GetDirectory(file.Folder.FullName).GetFileReference(file.Name)
        End Function

        Private Function GetFile(ByVal folderName As String, ByVal fileName As String) As CloudFile
            Return GetDirectory(folderName).GetFileReference(fileName)
        End Function

        Private Sub CopyFolderInternal(ByVal sourceFolder As CloudFileDirectory, ByVal targetFolder As CloudFileDirectory, ByVal directoryName As String)
            targetFolder.GetDirectoryReference(directoryName).CreateIfNotExists()
            Dim newFolder = targetFolder.GetDirectoryReference(directoryName)
            Dim items = sourceFolder.ListFilesAndDirectories()
            For Each item As IListFileItem In items
                If item.GetType() Is GetType(CloudFile) Then
                    Dim file = CType(item, CloudFile)
                    newFolder.GetFileReference(file.Name).StartCopy(file)
                Else
                    Dim directory = CType(item, CloudFileDirectory)
                    CopyFolderInternal(directory, newFolder, directory.Name)
                End If
            Next
        End Sub

        Private Sub CopyFolderInternal(ByVal sourceFolder As FileManagerFolder, ByVal newName As String)
            CopyFolderInternal(GetDirectory(sourceFolder.FullName), GetDirectory(sourceFolder.Parent.FullName), newName)
        End Sub

        Private Sub DeleteFolderInternal(ByVal sourceFolder As CloudFileDirectory)
            Dim items = sourceFolder.ListFilesAndDirectories().ToList()
            If items.Count = 0 Then
                sourceFolder.DeleteIfExists()
                Return
            End If

            For Each item As IListFileItem In items
                If item.GetType() Is GetType(CloudFile) Then
                    Dim file = CType(item, CloudFile)
                    file.DeleteIfExists()
                Else
                    Dim directory = CType(item, CloudFileDirectory)
                    DeleteFolderInternal(directory)
                End If
            Next

            sourceFolder.DeleteIfExists()
        End Sub
    End Class
End Namespace
